larq/models.py

```diff
@@ -47,82 +64,286 @@ def _get_output_shape(layer):
+
 
+class LayerProfile:
+    def __init__(self, layer):
+        self._layer = layer
+        self.parameter_profiles = []
+        self.op_profiles = []
 
-def _generate_table(model):
-    layer_stats = [_parse_params(l) for l in model.layers]
-    summed_stat = _sum_params(layer_stats)
+        for weight in layer.weights:
+            self.parameter_profiles.append(
+                ParameterProfile(weight, self._get_bitwidth(weight))
+            )
+
+        if type(layer) in mac_layers:
+            # mac layers should have a kernel weight and perhaps a bias
+            assert len(self.parameter_profiles) <= 2
+
+            for p in self.parameter_profiles:
+                if not p.is_bias():
+                    self.op_profiles.append(
+                        OperationProfile(
+                            n=p.count * self.output_pixels,
+                            precision=max(self.input_precision(32), p.bitwidth),
+                            op_type="mac",
+                        )
+                    )
+
+    @property
+    def memory(self):
+        return sum([p.memory for p in self.parameter_profiles])
+
+    @property
+    def fp_equivalent_memory(self):
+        return sum([p.fp_equivalent_memory for p in self.parameter_profiles])
+
+    def parameter_count(self, bitwidth=None, trainable=None):
+        count = 0
+        for p in self.parameter_profiles:
+            if bitwidth and (p.bitwidth != bitwidth):
+                continue
+            if (trainable is not None) and (p.trainable != trainable):
+                continue
+            count += p.count
+        return count
+
+    def op_count(self, precision=None, op_type=None, escape="?"):
+        if type(self._layer) in op_count_supported_layer_types:
+            count = 0
+            for op in self.op_profiles:
+                if precision and (op.precision != precision):
+                    continue
+                if op_type and (op.op_type != op_type):
+                    continue
+                count += op.n
+            return count
+        else:
+            return escape
+
+    def input_precision(self, default="-"):
+        try:
+            return self._layer.input_quantizer.precision
+        except:
+            return default
+
+    @property
+    def output_shape(self):
+        try:
+            return tuple(dim if dim else -1 for dim in self._layer.output_shape)
+        except AttributeError:
+            return "multiple"
+        except RuntimeError:  # output_shape unknown in Eager mode.
+            return "?"
+
+    @property
+    def output_pixels(self):
+        if len(self.output_shape) == 4:
+            return np.prod(self.output_shape[1:2])
+        elif len(self.output_shape) == 2:
+            return 1
+        else:
+            raise NotImplementedError()
+
+    @property
+    def unique_param_bidtwidths(self):
+        return sorted(np.unique([p.bitwidth for p in self.parameter_profiles]))
```

Comment:
Since we are sorting a pure Python `list`, converting it to a `set` is usually faster than using `np.unique` since this would need to convert it to an `np.ndarray` and then back to be used by `sorted`.
```suggestion
        return sorted(set([p.bitwidth for p in self.parameter_profiles]))
```